dotnet/src/SemanticKernel/Orchestration/Plan.cs

```diff
@@ -225,22 +314,193 @@ public async Task<Plan> InvokeNextStepAsync(SKContext context)
+            {
+                if (!keysToIgnore.Contains(key, StringComparer.InvariantCultureIgnoreCase) && functionVariables.Get(key, out var value))
+                {
+                    this.State.Set(key, value);
+                }
+            }
+
+            this.State.Update(result.Result.Trim());
+
+            foreach (var item in step.NamedOutputs)
+            {
+                if (string.IsNullOrEmpty(item.Key) || item.Key.ToUpperInvariant() == "INPUT" || string.IsNullOrEmpty(item.Value))
+                {
+                    continue;
+                }
+
+                if (item.Key.ToUpperInvariant() == "RESULT")
+                {
+                    this.State.Set(item.Value, result.Result.Trim());
+                }
+                else if (result.Variables.Get(item.Key, out var value))
+                {
+                    this.State.Set(item.Value, value);
+                }
+            }
+            // if (string.IsNullOrEmpty(sequentialPlan.OutputKey))
+            // {
+            //     _ = this.State.Update(result.Result.Trim());
+            // }
+            // else
+            // {
+            //     this.State.Set(sequentialPlan.OutputKey, result.Result.Trim());
+            // }
+
+            // _ = this.State.Update(result.Result.Trim());
+            // if (!string.IsNullOrEmpty(sequentialPlan.OutputKey))
+            // {
+            //     this.State.Set(sequentialPlan.OutputKey, result.Result.Trim());
+            // }
+
+            // if (!string.IsNullOrEmpty(sequentialPlan.ResultKey))
+            // {
+            //     _ = this.State.Get(SkillPlan.ResultKey, out var resultsSoFar);
+            //     this.State.Set(SkillPlan.ResultKey,
+            //         string.Join(Environment.NewLine + Environment.NewLine, resultsSoFar, result.Result.Trim()));
+            // }
+
+            #endregion Update State
+
             this.NextStepIndex++;
-            this.State.Update(context.Result.Trim());
         }
 
         return this;
     }
 
-    protected void SetFunction(ISKFunction function)
+    /// <summary>
+    /// Get JSON representation of the plan.
+    /// </summary>
+    public string ToJson()
+    {
+        return JsonSerializer.Serialize(this);
+    }
+
+    /// <summary>
+    /// To help with reading plans from <see cref="Orchestration.ContextVariables"/>.
+    /// </summary>
+    /// <param name="json">JSON string representation of aPlan</param>
+    /// <param name="context">The context to use for function registrations.</param>
+    /// <returns>An instance of a Plan object.</returns>
+    public static Plan FromJson(string json, SKContext context)
+    {
+        var plan = JsonSerializer.Deserialize<Plan>(json, new JsonSerializerOptions() { IncludeFields = true }) ?? new Plan(string.Empty);
+
+        plan = GetRegisteredFunctions(plan, context);
+
+        return plan;
+    }
+
+    private static Plan GetRegisteredFunctions(Plan plan, SKContext context)
```

Comment:
Would rename this function to `UpdateRegisteredFunctions()`
```suggestion
    private static Plan UpdateRegisteredFunctions(Plan plan, SKContext context)
```

Reply:
How about `SetRegisteredFunctions` since it's calling `SetFunction`?